enterprise/suppress-user/handler.go

```diff
@@ -0,0 +1,57 @@
+package suppression
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/utils/logger"
+	"github.com/samber/lo"
+)
+
+// newHandler creates a new handler for the suppression feature
+func newHandler(ctx context.Context, r Repository, log logger.Logger) *handler {
+	h := &handler{
+		r:   r,
+		log: log,
+	}
+	// we don't want to flood logs if, e.g. the suppression repository is restoring,
+	// so we are debouncing the logging
+	h.errLog.debounceLog, h.errLog.cancel = lo.NewDebounce(1*time.Second, func() {
+		h.errLog.errMu.Lock()
+		defer h.errLog.errMu.Unlock()
+		if h.errLog.err != nil {
+			h.log.Warn(h.errLog.err.Error())
+		}
+	})
+	go func() {
+		<-ctx.Done()
+		h.errLog.cancel()
+	}()
+	return h
+}
+
+// handler is a handle to this object
+type handler struct {
+	log    logger.Logger
+	r      Repository
+	errLog struct {
+		debounceLog func() // logs suppression failures with a debounce, once every 1 second
+		cancel      func() // cancels the debounce timer
+		errMu       sync.Mutex
+		err         error // the last error
+	}
+}
+
+func (h *handler) IsSuppressedUser(workspaceID, userID, sourceID string) bool {
+	h.log.Debugf("IsSuppressedUser called for workspace: %s, user %s, source %s", workspaceID, userID, sourceID)
+	suppressed, err := h.r.Suppress(workspaceID, userID, sourceID)
+	if err != nil {
+		h.errLog.errMu.Lock()
+		h.errLog.err = fmt.Errorf("suppression check failed for workspace: %s, user: %s, source: %s: %w", workspaceID, userID, sourceID, err)
+		h.errLog.debounceLog()
+		h.errLog.errMu.Unlock()
```

Comment:
I agree with you @saurav-malani, @fracasula wdyt?

Reply:
Can we use a channel instead with a leaky bucket approach? Some kind of variant of this:

```go
package main

import (
	"fmt"
	"time"
)

func main() {
	ch := make(chan error, 1)
	ticker := time.NewTicker(time.Second)
	timeout := time.After(5 * time.Second)

	go func() {
		for i := 0; ; i++ {
			select {
			case ch <- fmt.Errorf("error %d", i):
			default:
				// discarding error
			}
		}
	}()

	for {
		select {
		case <-ticker.C:
			fmt.Println(<-ch)
		case <-timeout:
			return
		}
	}
}
```

It would help us avoid both the race and the mutex. Thoughts?

Reply:
so keep the first error instead of the last one

Reply:
yeah but it gets refreshed even now and then, right? would that be good enough for this use case in your opinion?

Reply:
Alternative to keep last error:

```go
package main

import (
	"fmt"
	"time"
)

func main() {
	ch := make(chan error, 1)
	ticker := time.NewTicker(time.Second)
	timeout := time.After(5 * time.Second)

	go func() {
		for i := 0; ; i++ {
			ch <- fmt.Errorf("error %d", i)
		}
	}()

	go func() {
		var err error
		for {
			select {
			case err = <-ch:
			case <-ticker.C:
				fmt.Println(err)
			}
		}
	}()

	for {
		select {
		case <-timeout:
			return
		}
	}
}
```

`err` is local to that routine only and we either update err or log so it's race safe. wdyt of this?

Reply:
in what sense does this differ compared to using a mutex? Still the channel acts as a (blocking) synchronisation point

Reply:
There is a data race, I posted the full stack about it earlier. That will make tests with `-race` fail and additionally it might create other unknown issues (see previous post with C++ example). The channel can be buffered to give you enough room. A very small buffer (e.g. `1000`) might be good enough already considering that on the other side you would just `err = <-ch` which is a super fast operation. Way better than a mutex in my opinion and solves the data race (and you get rid of the debounce 3rd party dependency).

Reply:
Anyway you know my opinion by now :sweat_smile:
Unless you have a very very (**very**) good reason to ignore a data race you should not. Especially if you don't even know what kind of overhead you're paying to avoid it and whether that overhead is actually negligible.

In any case, maybe others can jump in and share their view.

Reply:
So I am removing the debounce logic as I now choose not to log any error while the repository is restoring. Any other error will be logged normally :)

Reply:
btw, we could avoid all the trouble by using an `atomic.Value` 🥴

Reply:
That would still block, right? Buffered channel should be the fastest race free solution (I guess) :thinking:

Reply:
What makes you think that a channel isn't blocking? At least it is using a mutex [internally](https://go.dev/src/runtime/chan.go). My experience with go channels so far is that channels are usually the less performant, but who knows!

Reply:
It makes me think so because **_a)_** `ch` can be buffered and **_b)_** on the other end you're just doing `err = <-ch` so I thought that it'd be faster because on the receiving side there is no blocking at all and on the publish side the buffer might be enough to have a non-blocking publish most of the time, at least in a low concurrency environment.

I wrote a quick benchmark and I got:
```
goos: linux
goarch: amd64
pkg: github.com/rudderlabs/rudder-server/utils/test
cpu: 12th Gen Intel(R) Core(TM) i9-12900K
Benchmark
Benchmark/mutex
    benchmark_test.go:30: error 0
    benchmark_test.go:30: error 99
    benchmark_test.go:30: error 7858
    benchmark_test.go:30: error 780155
    benchmark_test.go:30: error 1410642
Benchmark/mutex-24         	 1410706	       950.6 ns/op
Benchmark/channel
    benchmark_test.go:70: error 0
    benchmark_test.go:70: error 90
    benchmark_test.go:70: error 6384
    benchmark_test.go:70: error 999978
    benchmark_test.go:70: error 1897742
Benchmark/channel-24       	 1898244	       876.9 ns/op
PASS
```

For reference and in case I messed up with the code, please have a look to see if it makes sense:
```go
package test

import (
	"fmt"
	"sync"
	"testing"
)

func Benchmark(b *testing.B) {
	b.Run("mutex", func(b *testing.B) {
		var (
			err error
			mu  sync.Mutex
			wg  sync.WaitGroup
		)

		wg.Add(b.N)
		b.ResetTimer()
		for i := 0; i < b.N; i++ {
			go func(i int) {
				mu.Lock()
				defer mu.Unlock()
				err = fmt.Errorf("error %d", i)
				wg.Done()
			}(i)
		}

		wg.Wait()
		b.StopTimer()
		b.Log(err)
	})
	b.Run("channel", func(b *testing.B) {
		var (
			err  error
			wg   sync.WaitGroup
			ch   = make(chan error, 1000)
			done = make(chan struct{})
		)

		go func() {
			defer close(done)
			var (
				e    error
				open bool
			)
			for {
				select {
				case e, open = <-ch:
					if !open {
						return
					}
					err = e
				}
			}
		}()

		wg.Add(b.N)
		b.ResetTimer()
		for i := 0; i < b.N; i++ {
			go func(i int) {
				ch <- fmt.Errorf("error %d", i)
				wg.Done()
			}(i)
		}

		wg.Wait()
		close(ch)
		<-done
		b.StopTimer()
		b.Log(err)
	})
}
```

Reply:
FYI, with `ch   = make(chan error, b.N)` instead of `ch   = make(chan error, 1000)` I get:
```
goos: linux
goarch: amd64
pkg: github.com/rudderlabs/rudder-server/utils/test
cpu: 12th Gen Intel(R) Core(TM) i9-12900K
Benchmark
Benchmark/mutex
    benchmark_test.go:30: error 0
    benchmark_test.go:30: error 98
    benchmark_test.go:30: error 9766
    benchmark_test.go:30: error 996597
    benchmark_test.go:30: error 1065557
Benchmark/mutex-24         	 1340248	       881.4 ns/op
Benchmark/channel
    benchmark_test.go:70: error 0
    benchmark_test.go:70: error 98
    benchmark_test.go:70: error 9389
    benchmark_test.go:70: error 998271
    benchmark_test.go:70: error 3425483
Benchmark/channel-24       	 3426110	       340.9 ns/op
PASS
```

Interesting stuff.